sound/soc/sof/ipc4-topology.c

```diff
@@ -834,26 +835,52 @@ static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
 	if (ret)
 		goto err;
 
-	cfg_size = sizeof(struct sof_ipc4_base_module_cfg);
+	ret = sof_ipc4_widget_setup_msg(swidget, &process->msg);
+	if (ret)
+		goto err;
+
+	/* parse process init module payload config type from module info */
+	fw_module = swidget->module_info;
+	process->init_config = FIELD_GET(SOF_IPC4_MODULE_INIT_CONFIG_MASK,
+					 fw_module->man4_module_entry.type);
+
+	process->ipc_config_size = sizeof(struct sof_ipc4_base_module_cfg);
+
+	/* allocate memory for base config extension if needed */
+	if (process->init_config == SOF_IPC4_MODULE_INIT_CONFIG_TYPE_BASE_CFG_WITH_EXT) {
+		struct sof_ipc4_base_module_cfg_ext *base_cfg_ext;
+		u32 ext_size;
+
+		ext_size += struct_size(base_cfg_ext, pin_formats,
+					swidget->num_input_pins + swidget->num_output_pins);
 
-	cfg = kzalloc(cfg_size, GFP_KERNEL);
+		base_cfg_ext = kzalloc(ext_size, GFP_KERNEL);
+		if (!base_cfg_ext) {
+			ret = -ENOMEM;
+			goto free_available_fmt;
+		}
+
+		base_cfg_ext->num_input_pin_fmts = swidget->num_input_pins;
+		base_cfg_ext->num_output_pin_fmts = swidget->num_output_pins;
+		process->base_config_ext = base_cfg_ext;
+		process->base_config_ext_size = ext_size;
+		process->ipc_config_size += ext_size;
+	}
+
+	cfg = kzalloc(process->ipc_config_size, GFP_KERNEL);
 	if (!cfg) {
 		ret = -ENOMEM;
-		goto free_available_fmt;
+		goto free_bast_cfg_ext;
 	}
 
 	process->ipc_config_data = cfg;
-	process->ipc_config_size = cfg_size;
-	ret = sof_ipc4_widget_setup_msg(swidget, &process->msg);
-	if (ret)
-		goto free_cfg_data;
 
 	sof_ipc4_widget_update_kcontrol_module_id(swidget);
 
 	return 0;
-free_cfg_data:
-	kfree(process->ipc_config_data);
-	process->ipc_config_data = NULL;
```

Comment:
Where do we free `process->ipc_config_data` now?

Reply:
> Where do we free `process->ipc_config_data` now?

Please ignore this comment. `cfg` is allocated in the last step and no error handling after that.